torch/csrc/utils/python_dispatch.cpp

```diff
@@ -195,16 +177,25 @@ void initDispatchBindings(PyObject* module) {
           "impl",
           [](py::object self,
              const char* name,
-             const char* dispatch,
+             // TODO: empty string no longer works
+             c10::DispatchKey dispatch,
              py::object func) {
             HANDLE_TH_ERRORS
-            self.cast<torch::Library&>().impl(
+            auto& lib = self.cast<torch::Library&>();
+            lib.impl(
                 name,
-                dispatch_str(
+                torch::dispatch(
                     dispatch,
                     CppFunction::makeFromBoxedFunctor(
-                        std::make_unique<PythonKernelHolder>(
-                            std::move(func)))));
+                        std::make_unique<
+                            at::impl::PythonOpRegistrationTrampoline>(
+                            dispatch))));
+            // NB: this must happen AFTER the impl, because the OperatorHandle
+            // is not guaranteed to exist before we impl
+            python_registrations_[lib._resolve(name)].insert_or_assign(
```

Comment:
Hmm - it seems like the key thing is:
```
  // Which interpreter should we call?  Python operator registrations operate
  // in "hermetic mode", which means that tensors passing to Python get fresh
  // Python objects (not associated with the previous interpreter), and Python
  // objects aren't preserved when the Python objects go out of scope.  This
  // means that, in fact, these calls can be serviced by ANY interpreter.  So
  // we just arbitrarily pick one and jump to it.
```

Where if I'm following correctly, the "arbitrarily pick one and jump to it" happens when the trampoline code looks up the `SafePyObj` from this map (`python_registrations_`), and it's "arbitrary" because the values in this map can be overwritten multiple times when different PyInterpreters register the same op/key/function, so it's semi-arbitrary which one registers last and leaves their function in this map.

Instead of having a second out-of-band mapping from `operator` -> `dispatch_key` -> SafePyObj` that the trampoline needs to look up, is there any reason why we can't just directly store the current function/interpreter combo (the `SafePyObj` that we want to call) directly on the trampoline? That would save us from doing a bunch of extra map lookups, and keeping a separate map around. Let me know if I'm missing something though.

Reply:
Not quite. `python_registrations_` is defined in torch/csrc so it is only available from libtorch_python, but we need to make an interpreter decision in libtorch. The arbitrary interpreter is grabbed via `  static std::atomic<c10::impl::PyInterpreter*> interpreter_;` in `PythonOpRegistrationTrampoline`.

It may be possible to store the py object on the trampoline though. To do this, though, I will have to make it so that ONLY the victim interpreter actually does registrations, and everyone else whiffs them. I guess I will go implement this.

@d4l3k one thing this discussion clued me in on, is what if the victim interpreter gets unloaded before everything else is done? It seems like we will need to maintain an invariant that the first multipy interpreter always stays live.

Reply:
There is also an extra subtlety, which is that if I have two interpreters racing initialization, one interpreter could attempt to use an operator registration before the first interpreter had finished initializing all of its modules. One easy fix for this is to delay actually doing the CEX for making an interpreter visible until we're done processing `import torch`, but if there are Python operator registrations from other libraries, which could possibly be lazily imported, you may have to wait arbitrarily long for the first interpreter to be ready. So here I'm not so sure how multipy can even tell when all the ops are loaded.

Reply:
Would it help instead of using an arbitrary interpreter (depending on which one started up first), we set things up such that the "chosen" interpreter corresponds to the first interpreter to make an operator call?

Although... I'm assuming that if an interpreter is making an op call then all of it's ops have already been registered. I'm not sure if that's always the case.

And that would also have a cost, since we'd be pushing the "set up the interpreter" step from registration time to the time we invoke an operator.

Reply:
If you can assume that an operator is only called after everything is loaded, this should work. But if there can be operator calls and imports interleaved, this isn't sound. In fact I don't think there is anything we can do. So I'm just going to write a nice error message.